getting_started/scripting/gdscript/gdscript_styleguide.rst

```diff
@@ -3,21 +3,90 @@
 GDScript style guide
 ====================
 
-Description
------------
-
-This styleguide lists conventions to write elegant GDScript. The goal is
-to encourage writing clean, readable code and promote consistency across
-projects, discussions, and tutorials. Hopefully, this will also
-encourage development of auto-formatting tools.
+This style guide lists conventions to write elegant GDScript. The goal is to
+encourage writing clean, readable code and promote consistency across projects,
+discussions, and tutorials. Hopefully, this will also support the development of
+auto-formatting tools.
 
 Since GDScript is close to Python, this guide is inspired by Python's
 `PEP 8 <https://www.python.org/dev/peps/pep-0008/>`__ programming
-styleguide.
+style guide.
+
+Style guides aren't meant as hard rulebooks. At times, you may not be able to
+apply some of the guidelines below. When that happens, use your best judgment,
+and ask fellow developers for insights.
+
+In general, keeping your code consistent in your projects and within your team is
+more important than following this guide to a tee.
 
 .. note:: Godot's built-in script editor uses a lot of these conventions
           by default. Let it help you.
 
+Here is a complete class example based on these guidelines:
+
+::
+
+    class_name StateMachine
+    extends Node
+    # Hierarchical State machine for the player.
+    # Initializes states and delegates engine callbacks (_physics_process, _unhandled_input) to the state.
+
+
+    signal state_changed(previous, new)
+
+    export var initial_state = NodePath()
+    var is_active = true setget set_is_active
+
+    onready var state = get_node(initial_state) setget set_state
```

Comment:
in the past, I had a discussion with @TheDuriel regarding `public onready variable` and the outcome was that it is always bad practice since it's all about exposing internals to the clients

Reply:
So what would you do with this case? You'd encapsulate the states in a state machine away from the users? And... add extra methods and wrapper properties to get the info you need from them?

If it's public here, it's because we want to access the active state from outside the state machine. Among other things, we've used the property to display info about the active state on-screen, in a debug console.

In this case, it's quite inconvenient to e.g. bubble the state with signals. And the above code is not inconsistent with some of Godot's API, which exposes some members nodes or objects as properties at times.

Reply:
I can make the state variable private no problem. Making it public is not necessarily the best example for e.g. a dev who doesn't have much experience and who may use that in a different context.

Reply:
Also if you'd prefer another class example I can come up with one - it's just that this is a complete class that is short and shows most of the guidelines at the same time.

Reply:
either way will be fine ^^